Make required actions configurable (#28400)

thomasdarimont opened the pull request:
This PR adds support for configurable required actions.

Some implementation notes:

- Introduced `RequiredActionConfigModel` to hold the configration of a RequiredAction
- Extended `AuthenticationManagementResource` with CRUD operations for RequiredAction config management
- Extended admin-client libraries with RequiredAction config management methods
- Exposed `RequiredActionConfigModel` via RequiredActionContext to users
- Add tests for crud operations on configurable required actions

Fixes #28400

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
Make required actions configurable (#28400)

- Add tests for crud operations on configurable required actions
- Add support exposing the required action configuration via RequiredActionContext
- Make configSaveError message reusable in other contexts
- Introduced admin-ui specific endpoint for retrieving required actions with config metadata

Fixes #28400

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>
Co-authored-by: Erik Jan de Wit <erikjan.dewit@gmail.com>

thomasdarimont commented:
@stianst this is the PR for initial support for configurable required actions.

What's already working:
- Storage of required actions configuration in the `required_action_config` table
- Configuration per required actions instance in a realm
- CRUD for required actions with configuration
- Users can make their own required action implementations configurable by overriding the `org.keycloak.provider.ProviderFactory#getConfigMetadata()`
- Initial tests for RequiredAction CRUD config management
- Extended admin-client libraries

What's still missing:
- Admin UI integration ("cog" settings button and settings form for required actions)
- Documentation
- "Proper" way to validate config settings before save

thomasdarimont reviewed (commented):
This is just a placeholder for the configure button

thomasdarimont commented on js/apps/admin-ui/src/authentication/RequiredActions.tsx:
```diff
@@ -133,6 +136,18 @@ export const RequiredActions = () => {
           displayKey: "requiredActions",
         },
+        {
+          name: "config",
```
This is just a placeholder for the "configure" button for required actions.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/authentication/RequiredActionFactory.java:
```diff
@@ -44,4 +47,24 @@ public interface RequiredActionFactory extends ProviderFactory<RequiredActionPro
+     * @param realm
+     * @param model
+     */
+    default void validateConfig(RealmModel realm, RequiredActionConfigModel model) {
```
I'm not sure whether this is the proper way to perform validation, but it could be used to prevent bad configurations to be written to the database.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/utils/CredentialHelper.java:
```diff
@@ -90,6 +95,33 @@ public static ConfigurableAuthenticatorFactory getConfigurableAuthenticatorFacto
          return factory;
     }
 
+    public static RequiredActionFactory getConfigurableRequiredActionFactory(KeycloakSession session, String providerId) {
```
This and the following helper functions should be moved elsewhere... just placed them here because the authenticator config methods were also located here.

mposolda replied:
Nitpick: Can be ok to introduce new class like `RequiredActionHelper` and add those methods here as they don't have much to do with credentials...

thomasdarimont reviewed (commented):
Some remarks inline.

thomasdarimont commented:
I now have an initial version working, that allows to inspect, add, update and remove required action configuration via the admin ui.

Required Actions List:
I now only show the settings icon, if the required action is configurable. Instead of using a dedicated column for the button, we could also place the icon next to the required action label.
![image](https://github.com/keycloak/keycloak/assets/314690/014cdaf1-f8be-4fd0-b7de-2cd3e64f0d45)

Required Actions Config Modal:
In the config modal I currently only show the actual settings. It would be good to show the actual name of the current required action as a label somewhere.
![image](https://github.com/keycloak/keycloak/assets/314690/3eb12ee5-883f-4141-9287-2d7180dc61a9)

thomasdarimont reviewed (commented):
Added some more remarks to the places where I'm unsure

thomasdarimont commented:
Thanks to @edewit the new UI now looks a bit better:
![image](https://github.com/keycloak/keycloak/assets/314690/73c72188-a3f7-44c0-baac-18cda8024934)

Config screen:
![image](https://github.com/keycloak/keycloak/assets/314690/9bda21af-e1d6-4462-8836-49d2d4658a9d)

@edewit would it be possible to show the translated name instead of the alias? (Update Password instead of "UPDATE_PASSWORD"). Also I think the config dialog would scale better if the name of the required action would be on a separat line. Some users tend to have quite long (but descriptive) names for required actions.

mposolda commented on server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java:
```diff
@@ -120,4 +123,12 @@ public Map<String, String> getConfig() {
         this.config = config;
     }
+
+    public boolean isConfigurable() {
```
Do we need `isConfigurable` and `setConfigurable` on the model? It seems that the `configurable` field is not backed in the DB and it is always derived based on the factory (like in `RealmAdapter.entityToModel(RequiredActionProviderEntity entity)`).

Also I believe it is not needed in the JSON representation object, which is used for export/import, but I understand it can be useful for have this in the UI layer. Is it possible to decouple those?

thomasdarimont replied:
Thanks for taking a look! Yes I mostly added the configurable property to the `RequiredActionProviderModel` and `RequiredActionProviderRepresentation` . Initially I thought about deriving the "configurability" of a RequiredAction from a non-empty configuration. However, with this we currently cannot differentiate an empty configuration from a non-configurable required action, as the config propery is currently always initialized with an empty map here: https://github.com/keycloak/keycloak/blob/main/core/src/main/java/org/keycloak/representations/idm/RequiredActionProviderRepresentation.java#L35
If we would change this to be `null` by default, then I could remove the `configurable` from the model and the representation property. 

I was also thinking about using the component model but then realized that required actions already have a database backed config HashMap, so I kept using that.

mposolda replied:
I understand that the info about "configurable" is needed in the UI. But I think it is not needed in the model layer and might be cleaner to remove it from here.

I see the options like:

1) Keep `configurable` at the model layer and `RequiredActionProviderRepresentation`. Not so clean IMO.
2) Use the `null` to indicate. It can work, but just afraid a bit of things like `NullPointerException` etc.
3) Add stuff inside admin UI java extension (module `rest/admin-ui-ext` in the workspace) and make admin console to use it. I can imagine something like this in the package `org.keycloak.admin.ui.rest.model` inside admin UI extension (the stuff in the module `rest`, which contains REST endpoints used solely by admin console):

```
public class ConfigurableRequiredActionProviderRepresentation extends RequiredActionProviderRepresentation {

    private boolean configurable;

    // Getter and setter for "configurable" here ...
}
```

And then make `org.keycloak.admin.ui.rest.AuthenticationManagementResource` to return required actions like those, so the admin console UI is able to figure if action is configurable or not. The disadvantage of the extension is additional code.

My vote is either (2) or (3). WDYT?

mposolda reviewed (commented):
Thanks,

I've added one minor comment inline. In general, I am not 100% sure if we want to rely on `ComponentModel`, which was the approach discussed on the meeting AFAIR. But maybe not... (added some more details in the email).

thomasdarimont commented:
A quick experiment for better validation support: 
![image](https://github.com/keycloak/keycloak/assets/314690/f1ed568a-4825-44e7-be59-208d23ba6619)

thomasdarimont commented:
Now using the required action `name` instead of `alias`, as this is more user friendly.
![image](https://github.com/keycloak/keycloak/assets/314690/4ff50ec8-780f-41cd-9610-b54ef50dfc63)

thomasdarimont commented:
@mposolda thanks for your review and the helpful suggestions! Regarding your proposal for moving forward, I went with a combination of (2) and (3) as discussued above:
- (2) use null to indicate no required action config is set
- (3) add  admin ui specific endpoint  that returns the `configurable` property for a required action.

The latest changes of this PR include:
- Removed boolean configurable from RequiredActionProviderRepresentation and RequiredActionConfigModel classes
- Adapted UI and API to work with requiredaction provider alias instead of providerId for consistency
- Introduced admin-ui specific endpoint org.keycloak.admin.ui.rest.AuthenticationManagementResource#getRequiredActions for retrieving required actions with config metadata

Since all the endpoints that handle requiredactions work with the requiredaction `alias` value,
I adapted all places that work with requiredaction configs to use aliases as well.
When I have the `providerId`, then I also check if the `providerId` matches that of the given `RequiredActionConfigModel`.

Only difference is, which only has an alias available at the moment:
org.keycloak.models.jpa.RealmAdapter#getRequiredActionConfigByAlias

However, using the aliases all the time means that required action aliases within a realm must be unique - is this guaranteed somehow?

I also tried to have `max_auth_age` as a generic configuration parameter for all required actions, but I needed to rollback that change I wasn't sure whether this really makes sense for all required actions and wanted to discuss this first. 

As an example I adapted the `UPDATE_PASSWORD` required action to consider the configured value for `max_auth_age` with a preference for the password policy if set.

I think the PR - at least for the initial capacity is on the finish line. Perhaps @edewit could take a look again on the translation of error messages if a validation fails :) (I need some help to render the proper error message based on the server response).
![image](https://github.com/keycloak/keycloak/assets/314690/17af88d4-4cb6-4113-ae25-de0c6c57b44f)

thomasdarimont commented:
When we allow configuration of required actions, it might make sense to add support for help texts that can describe the action and options.

I experimented with this a bit here:
![image](https://github.com/keycloak/keycloak/assets/314690/b409cc36-3728-4da3-99aa-14dba571949d)

Adding helpText support for required actions was a brief experiment - I'll leave it out from this PR and add it in a separate PR.

thomasdarimont commented on js/apps/admin-ui/src/authentication/RequiredActions.tsx:
```diff
@@ -65,6 +89,8 @@ export const RequiredActions = () => {
       if (field in action) {
         action[field] = !action[field];
+        // remove configurable property from action which only exists for the admin ui
+        delete action.configurable;
```
We need to omit the `configurable` property here. This is necessary since we use the configurable action representation in the admin UI with the `configurable` property, which currently does not exist on the required action representation used by the standard admin rest API.

thomasdarimont commented:
@mposolda I squashed the commits into one and rebased on current main.

ahus1 commented:
@thomasdarimont - there is another conflict, and it needs rebasing.

thomasdarimont commented:
@ahus1 I just rebased the code on latest main and resolved the conflicts.

thomasdarimont commented:
@jonkoops This PR currently fails due to an unrelated test:
```
  Running:  realm_settings_user_profile_enabled.spec.ts                                     (7 of 7)

  User profile tabs
    Attributes sub tab tests
      ✓ Goes to create attribute page (4828ms)
      ✓ Completes new attribute form and performs cancel (2892ms)
      ✓ Completes new attribute form and performs submit (2619ms)
      ✓ Modifies existing attribute and performs save (3569ms)
      ✓ Adds and removes validator to/from existing attribute and performs save (3593ms)
    Attribute groups sub tab tests 
      (Attempt 1 of 4) Deletes an attributes group  <---- This fails
      (Attempt 2 of 4) Deletes an attributes group
      (Attempt 3 of 4) Deletes an attributes group
      1) Deletes an attributes group
```

I ran the cypress tests locally and there it fails too. I currently don't see a link to my changes. Do you have an idea what the problem is?

jonkoops commented:
Not sure what is happening there, perhaps someone from the UI team can take a look. @edewit @hmlnarik

hmlnarik commented:
Looks like a test interference. I have introduced #29507 and rerun the admin UI testsuite from `Generate test seed` on.

thomasdarimont commented:
Unfortunately I'm blocked by the failing UI tests, which are unrelated to this PR. The same tests also fail on my machine locally when I run the latest Keycloak main (I tested this with cypress on electron, and chrome).
The ui test that currently fails is: `client_authorization_test.spec.ts`.

Any help would be highly appreciated :)

edewit commented:
> Any help would be highly appreciated :)

Tests have been fixed on main, I have rebased your PR

mposolda commented on server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java:
```diff
@@ -43,6 +45,7 @@ public interface RequiredActionComparator extends Comparator<RequiredActionProvi
     private boolean defaultAction;
     private int priority;
     private Map<String, String> config = new HashMap<>();
+    private boolean configurable;
```
Nitpick: This field is not used and can be removed from this class (probably all the changes from this class can be removed from this PR)

mposolda commented on services/src/main/java/org/keycloak/authentication/requiredactions/UpdatePassword.java:
```diff
@@ -205,12 +239,62 @@ public int getMaxAuthAge() {
+
+            // we need to figure out the alias for the current required action
+            String providerId = authSession.getClientNote(Constants.KC_ACTION);
+            RequiredActionProviderModel requiredAction = realm.getRequiredActionProvidersStream() //
```
Nitpick: I can imagine it can be useful to have some util method (maybe in the `RequiredActionsHelper` as suggested in other comments or in `KeycloakModelUtils`) like:
```
RequiredActionConfigModel getRequiredActionConfig(RealmModel realm, String requiredActionProviderId)
```
as I suppose the similar code will apply always when retrieving configurations from required action

mposolda commented on services/src/main/java/org/keycloak/authentication/requiredactions/UpdatePassword.java:
```diff
@@ -29,27 +35,55 @@
+                .property() //
+                .name(MAX_AUTH_AGE_KEY) //
+                .label("Maximum Age of Authentication") //
+                .helpText("Configures the duration in seconds this action can be used after the last authentication before the user is required to re-authenticate.") //
```
```suggestion
                .helpText("Configures the duration in seconds this action can be used after the last authentication before the user is required to re-authenticate. This parameter is used just in the context of AIA when the kc_action parameter is available in the request, which is for instance when user himself updates his password in the account console. When the 'Maximum Authentication Age' password policy is used in the realm, it's value has precedence over the value configured here.") //
```

Nitpick: just suggesting to add a bit more info to the helpText to clarify it is used just in the context of AIA and about the fact that password policy has preference.

mposolda reviewed (commented):
@thomasdarimont Thanks for all the updates!

I've added few comments inline. It can be good to address them, but at the same time, I don't consider any of them as a blocker, so we can possibly do as a follow-up. Whatever you prefer.

Will need review from UI team as well.

mposolda commented:
@thomasdarimont Since there is conflict and this PR will require changes, WDYT about adding my "nitpicks" (if they makes sense for you) when you work on resolving conflict? :-) Thanks!

thomasdarimont commented:
@mposolda Thanks for your review and the recommendations. I fixed all your nitpicks and rebased to latest main. 
I think the PR is now ready for merge.

mposolda reviewed (approved):
@thomasdarimont @edewit Thanks a lot for all the updates and review!
